server/src/main/java/io/crate/expression/scalar/ArrayPositionFunction.java

```diff
@@ -0,0 +1,121 @@
+package io.crate.expression.scalar;
+
+import static io.crate.expression.scalar.array.ArrayArgumentValidators.ensureInnerTypeIsNotUndefined;
+import static io.crate.metadata.functions.TypeVariableConstraint.typeVariable;
+import static io.crate.types.TypeSignature.parseTypeSignature;
+
+import java.util.List;
+import java.util.Objects;
+
+import io.crate.data.Input;
+import io.crate.metadata.NodeContext;
+import io.crate.metadata.Scalar;
+import io.crate.metadata.TransactionContext;
+import io.crate.metadata.functions.Signature;
+import io.crate.types.DataType;
+import io.crate.types.DataTypes;
+
+public class ArrayPositionFunction<T> extends Scalar<Integer, List<T>> {
+
+    public static final String NAME = "array_position";
+    private final DataType returnType;
+    private final Signature signature;
+    private final Signature boundSignature;
+
+    public ArrayPositionFunction(Signature signature, Signature boundSignature) {
+        this.signature = signature;
+        this.boundSignature = boundSignature;
+        this.returnType = signature.getReturnType().createType();
+        ensureInnerTypeIsNotUndefined(boundSignature.getArgumentDataTypes(), signature.getName().name());
+    }
+
+    public static void register(ScalarFunctionModule scalarFunctionModule) {
+        scalarFunctionModule.register(
+            Signature.scalar(NAME,
+                parseTypeSignature("array(T)"),
+                parseTypeSignature("T"),
+                DataTypes.INTEGER.getTypeSignature()
+            )
+                .withTypeVariableConstraints(typeVariable("T"))
+                .withForbiddenCoercion()
```

Comment:
Thanks a lot @seut for review comments. I'll fix those asap. The reason behind forbidden coercion is to prevent searching of heterogenous types in the array. I thought, searching for text in an integer array wouldn't makes sense so the search parameter type should be same as inner type of the array. This also simply saves me from putting validation logic for the types inside the evaluate method business logic. 

Please help me understand the requirement here. If there is some other way to do it then yes we don't need coercion forbidden as well as those many signatures.

Reply:
When using a type variable constraint inside the signature, like you've defined by `.withTypeVariableConstraints(typeVariable("T"))`, the signature matching logic ensures that the types must match. 
With `.withForbiddenCoercion()` any implicit cast is prevented (CrateDB uses type precedence for that), which may useful for example when different function implementations for different types are needed/wanted.

Reply:
Ok. Yes I did notice that implicit casting but for this function I choose to prevent that. So just to understand from few example. Let's say we have these calls
array_position([3,1,2,4,5,6] :: array(Integer) , 2.0) -> should return 3 as the target value is effectively equal? 

and what about this one
array_position([3,1,2,4,5,6] :: array(Integer) , '2') -> should return 3 as the target value is effectively equal? 

Although, I'll look deeper into it and bring some leniency to the function implementation w.r.t implicit cast.

Reply:
> array_position([3,1,2,4,5,6] :: array(Integer) , 2.0) -> should return 3 as the target value is effectively equal?

Yes. Due to type precedence rules, the array will be implicit cast to `array(double)` in that case.

> array_position([3,1,2,4,5,6] :: array(Integer) , '2') -> should return 3 as the target value is effectively equal?

Yes. All these cases work as long as implicit cast rules are sufficient for it. In that case, the string literal must be castable to an integer, providing e.g. 'foo' will fail with a cast error.

Reply:
Forbidden coercion has been removed.